### PR TITLE
[ARCTIC-459][FLINK] Fix the missing of the required fields after project pushdown

### LINKED ISSUE #459: [Bug][Flink]: Project pushdown may lead to the missing of primary key, partition key, watermark field.
### What happened?

Project pushdown may lead to the missing of primary key, partition key, watermark field.  ShuffleHelper and watermark generator may lose the required info, and produce NPE.

### Affects Versions

0.3.2

### What engines are you seeing the problem on?

_No response_

### How to reproduce

create table with partition fileds day, name and normal fields id, age.
select id, age from table.

### Relevant log output

```shell
java.lang.NullPointerException
	at com.netease.arctic.data.PrimaryKeyData.<init>(PrimaryKeyData.java:57)
	at com.netease.arctic.flink.shuffle.ShuffleHelper.build(ShuffleHelper.java:62)
	at com.netease.arctic.flink.table.ArcticDynamicSource.hash(ArcticDynamicSource.java:205)
	at com.netease.arctic.flink.table.ArcticDynamicSource.distribute(ArcticDynamicSource.java:198)
	at com.netease.arctic.flink.table.ArcticDynamicSource.access$000(ArcticDynamicSource.java:67)
	at com.netease.arctic.flink.table.ArcticDynamicSource$1.produceDataStream(ArcticDynamicSource.java:141)
	at org.apache.flink.table.planner.plan.nodes.common.CommonPhysicalTableSourceScan.createSourceTransformation(CommonPhysicalTableSourceScan.scala:88)
	at org.apache.flink.table.planner.plan.nodes.physical.stream.StreamExecTableSourceScan.translateToPlanInternal(StreamExecTableSourceScan.scala:91)
	at org.apache.flink.table.planner.plan.nodes.physical.stream.StreamExecTableSourceScan.translateToPlanInternal(StreamExecTableSourceScan.scala:44)
	at org.apache.flink.table.planner.plan.nodes.exec.ExecNode.translateToPlan(ExecNode.scala:59)
	at org.apache.flink.table.planner.plan.nodes.exec.ExecNode.translateToPlan$(ExecNode.scala:57)
	at org.apache.flink.table.planner.plan.nodes.physical.stream.StreamExecTableSourceScan.translateToPlan(StreamExecTableSourceScan.scala:44)
	at org.apache.flink.table.planner.plan.nodes.physical.stream.StreamExecCalc.translateToPlanInternal(StreamExecCalc.scala:54)
	at org.apache.flink.table.planner.plan.nodes.physical.stream.StreamExecCalc.translateToPlanInternal(StreamExecCalc.scala:39)
	at org.apache.flink.table.planner.plan.nodes.exec.ExecNode.translateToPlan(ExecNode.scala:59)
	at org.apache.flink.table.planner.plan.nodes.exec.ExecNode.translateToPlan$(ExecNode.scala:57)
	at org.apache.flink.table.planner.plan.nodes.physical.stream.StreamExecCalcBase.translateToPlan(StreamExecCalcBase.scala:38)
	at org.apache.flink.table.planner.plan.nodes.physical.stream.StreamExecSink.translateToPlanInternal(StreamExecSink.scala:79)
	at org.apache.flink.table.planner.plan.nodes.physical.stream.StreamExecSink.translateToPlanInternal(StreamExecSink.scala:43)
	at org.apache.flink.table.planner.plan.nodes.exec.ExecNode.translateToPlan(ExecNode.scala:59)
	at org.apache.flink.table.planner.plan.nodes.exec.ExecNode.translateToPlan$(ExecNode.scala:57)
	at org.apache.flink.table.planner.plan.nodes.physical.stream.StreamExecSink.translateToPlan(StreamExecSink.scala:43)
	at org.apache.flink.table.planner.delegation.StreamPlanner.$anonfun$translateToPlan$1(StreamPlanner.scala:66)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:233)
	at scala.collection.Iterator.foreach(Iterator.scala:937)
	at scala.collection.Iterator.foreach$(Iterator.scala:937)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1425)
	at scala.collection.IterableLike.foreach(IterableLike.scala:70)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:69)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at scala.collection.TraversableLike.map(TraversableLike.scala:233)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:226)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at org.apache.flink.table.planner.delegation.StreamPlanner.translateToPlan(StreamPlanner.scala:65)
	at org.apache.flink.table.planner.delegation.PlannerBase.translate(PlannerBase.scala:162)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.translate(TableEnvironmentImpl.java:1329)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeInternal(TableEnvironmentImpl.java:676)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeOperation(TableEnvironmentImpl.java:767)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeSql(TableEnvironmentImpl.java:666)
	at com.netease.arctic.flink.FlinkTestBase.exec(FlinkTestBase.java:229)
	at com.netease.arctic.flink.FlinkTestBase.exec(FlinkTestBase.java:225)
	at com.netease.arctic.flink.table.TestJoin.t(TestJoin.java:175)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
```

### Anything else

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by zstraw**
FIX #459 

## Why are the changes needed?
Project pushdown may lead to the missing of primary key, partition key. ShuffleHelper may lose the required info, and produce NPE.

This hotfix tends to read the required fields appended after the projected fields in the order of 'projected fields', 'primary keys', 'partition keys'. And send the rowData with additional fields to downstream.

## Brief change log

  - *The flink 1.12 and 1.14, 1.15 versions are affected.*

## How was this patch tested?
- [X] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [X] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java [hunk 1]
```diff
@@ -121,4 +138,77 @@ public static RowType toRowType(TableSchema tableSchema) {
     }
     return RowType.of(fields);
   }
+
+  /**
+   * Primary key and partition key are the required fields for shuffle.
+   * The required fields should be added even though project push-down
```
Comment (SteNicholas) [inline flink/v1.12/flink/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java:144]:
```suggestion
   * The required fields should be added even though projection push down.
```

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java [hunk 2]
```diff
@@ -121,4 +138,77 @@ public static RowType toRowType(TableSchema tableSchema) {
...
+   */
+  public static List<Types.NestedField> addPrimaryAndPartitionKey(
+      List<Types.NestedField> projectedColumns, ArcticTable table) {
+    List<String> primaryKeys = table.isUnkeyedTable() ? Collections.EMPTY_LIST
+        : table.asKeyedTable().primaryKeySpec().fields().stream()
+        .map(PrimaryKeySpec.PrimaryKeyField::fieldName).collect(Collectors.toList());
+
+    List<Types.NestedField> columns = new ArrayList<>(projectedColumns);
+    Set<String> projectedNames = new HashSet<>();
```
Comment (SteNicholas) [inline flink/v1.12/flink/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java:153]: Simplied with Map<String, Integer> projections?

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java [hunk 3]
```diff
@@ -121,4 +138,77 @@ public static RowType toRowType(TableSchema tableSchema) {
...
+    for (PartitionField field : table.spec().fields()) {
+      partitionNames.add(field.name());
+    }
+
+    partitionNames.forEach(p -> {
+      if (projectedNames.contains(p)) {
+        return;
+      }
+      builder.field(p, tableSchema.getFieldDataType(p)
+          .orElseThrow(() -> new ValidationException("Arctic partition key should be declared in table")));
```
Comment (SteNicholas) [inline flink/v1.12/flink/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java:211]: Is there any case that Arctic primary key isn't declared in table here?

File: flink/v1.12/flink/src/test/java/com/netease/arctic/flink/table/TestWatermark.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    super.config(TEST_CATALOG_NAME);
+  }
+
+  @After
+  public void after() {
+    sql("DROP TABLE IF EXISTS arcticCatalog." + DB + "." + TABLE);
+  }
+
+  @Test
+  public void testWatermark() throws Exception {
```
Comment (SteNicholas) [inline flink/v1.12/flink/src/test/java/com/netease/arctic/flink/table/TestWatermark.java:79]: Could this create two test cases which test the watermark without projection and the watermark with projection?

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java [hunk 5]
```diff
@@ -121,4 +138,77 @@ public static RowType toRowType(TableSchema tableSchema) {
...
+    Set<String> projectedNames = Arrays.stream(projectedColumns).collect(Collectors.toSet());
+
+    if (table.isKeyedTable()) {
+      List<String> pks = table.asKeyedTable().primaryKeySpec().fieldNames();
+      pks.forEach(pk -> {
+        if (projectedNames.contains(pk)) {
+          return;
+        }
+        builder.field(pk, tableSchema.getFieldDataType(pk)
+            .orElseThrow(() -> new ValidationException("Arctic primary key should be declared in table")));
```
[the remark this answers is not in the document]
Comment (zstraw) [reply flink/v1.12/flink/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java:197]: > Is there any case that Arctic primary key isn't declared in table here?

To support "create table like"，Arctic connector supports create multi flink table with same underlying physical table like "create table XX (...) WITH ('connector'='arctic')".

So the primary key may not be declared in this case. But now this usage is only applied in 'create table like' and not recommended, there are no docs about that.

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/shuffle/ShuffleHelper.java [hunk 6]
```diff
@@ -66,7 +68,8 @@ public static ShuffleHelper build(ArcticTable table, Schema schema, RowType rowT
 
   /**
    * If using arctic table as build table, there will be an additional implicit field, valuing process time.
-   * @param schema The physical schema in Arctic table
+   *
+   * @param schema  The physical schema in Arctic table
```
Comment (SteNicholas) [inline flink/v1.12/flink/src/main/java/com/netease/arctic/flink/shuffle/ShuffleHelper.java:72]:
```suggestion
   * @param schema  The physical schema in Arctic table.
```

**Review by SteNicholas — APPROVED** [review]
LGTM.

**Review by zhoujinsong — APPROVED** [review]
LGTM.

**Review by YesOrNo828 — APPROVED** [review]

**Pull request merged**